community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/log/PhysicalLogFileRotateAndReadRaceIT.java

```diff
@@ -0,0 +1,152 @@
+ * There was this race where the log version was incremented, the new log file created and reader would get
+ * to this new file and try to read the header and fail before the header had been written.
+ *
+ * This test tries to reproduce this race. It will not produce false negatives, but sometimes false positives
+ * since it's non-deterministic.
+ */
+public class PhysicalLogFileRotateAndReadRaceIT
+{
+    private final TestDirectory directory = TargetDirectory.testDirForTest( getClass() );
+    private final LifeRule life = new LifeRule( true );
+    private final OtherThreadRule<Void> t2 = new OtherThreadRule<>( getClass().getName() + "-T2" );
+    @Rule
+    public final RuleChain rules = RuleChain.outerRule( directory ).around( life ).around( t2 );
+
+    // If any of these limits are reached the test ends, that or if there's a failure of course
+    private static final long LIMIT_TIME = SECONDS.toMillis( 5 );
+    private static final int LIMIT_ROTATIONS = 500;
+    private static final int LIMIT_READS = 1_000;
+
+    @Test
+    public void shouldNotSeeEmptyLogFileWhenReadingTransactionStream() throws Exception
+    {
+        // GIVEN
+        FileSystemAbstraction fileSystem = new DefaultFileSystemAbstraction();
+        PhysicalLogFiles logFiles = new PhysicalLogFiles( directory.directory(), fileSystem );
+        LogVersionRepository logVersionRepository = new DeadSimpleLogVersionRepository( 0 );
+        PhysicalLogFile.Monitor monitor = mock( PhysicalLogFile.Monitor.class );
+        LogHeaderCache headerCache = new LogHeaderCache( 10 );
+        PhysicalLogFile logFile = life.add( new PhysicalLogFile( fileSystem, logFiles, kibiBytes( 1 ),
+                () -> 2L, logVersionRepository, monitor, headerCache ) );
+        FlushablePositionAwareChannel writer = logFile.getWriter();
+        LogPositionMarker startPosition = new LogPositionMarker();
+        writer.getCurrentPosition( startPosition );
+
+        // WHEN
+        AtomicBoolean end = new AtomicBoolean();
+        byte[] dataChunk = new byte[100];
+        // one thread constantly writing to and rotating the channel
+        AtomicInteger rotations = new AtomicInteger();
+        Future<Void> writeFuture = t2.execute( ignored ->
+        {
+            ThreadLocalRandom random = ThreadLocalRandom.current();
+            while ( !end.get() )
+            {
+                writer.put( dataChunk, random.nextInt( 1, dataChunk.length ) );
+                if ( logFile.rotationNeeded() )
+                {
+                    logFile.rotate();
+                    // Let's just close the gap to the reader so that it gets closer to the "hot zone"
+                    // where the rotation happens.
+                    writer.getCurrentPosition( startPosition );
+                    rotations.incrementAndGet();
+                }
+            }
+            return null;
+        } );
```

Comment:
Perhaps wait for the `writeFuture` job to have started, by letting it release a latch of some kind as the first thing it does, and then let the main thread wait on that latch.